### PR TITLE
Fix unexpected IllegalReferenceCountException on decode multipart request

### PR DESCRIPTION
**Pull request by ssserj**
Fix unexpected IllegalReferenceCountException on decode multipart request. (#8575)

Motivation:

Http post request may be encoded as 'multipart/form-data' without any files and consist mixed attributes only.

Modifications:

Check reference count > 0 added on every InterfaceHttpData before release() call.

Result:

Code does not throw an IllegalReferenceCountException.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@ssserj imho that is not a correct fix. You should never need to check the refCnt before you release. This will just mask the bug. There must be something else wrong.

**Comment by kachayev** [comment]
@normanmaurer Unfortunately, decoder already contains code like this [here](https://github.com/netty/netty/blob/948d4a9ec58aef092c84eb76e672fd06e81cc13c/codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java#L940). I don't know why do we need to check `refCnt` there too.

**Comment by ssserj** [comment]
@normanmaurer The problem might be introduced by the commit 2b4f667. More detailed in issue #8575 and PR #8964

**Comment by normanmaurer** [comment]
@ssserj sorry for the long turn-around. Totally missed this one. Can you please sign our ICLA and let me know one done ?

https://netty.io/s/icla ?

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/HttpPostRequestDecoderTest.java [hunk 1]
```diff
@@ -656,4 +656,36 @@ public void testDecodeMalformedEmptyContentTypeFieldParameters() throws Exceptio
...
+                        java.nio.charset.Charset.forName("utf-8"));
+
+        assertTrue(decoder.isMultipart());
+        assertFalse(decoder.getBodyHttpDatas().isEmpty());
+        assertEquals(2, decoder.getBodyHttpDatas().size());
+        assertEquals("test message", ((Attribute) decoder.getBodyHttpData("msg")).getValue());
+        assertEquals("15200", ((Attribute) decoder.getBodyHttpData("msg_id")).getValue());
+
+        decoder.destroy();
+    }
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/HttpPostRequestDecoderTest.java]: can we also add a `assertEquals(0, req.refCnt());` after the `destroy()` call ?

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/HttpPostRequestDecoderTest.java [hunk 2]
```diff
@@ -656,4 +656,36 @@ public void testDecodeMalformedEmptyContentTypeFieldParameters() throws Exceptio
...
+                "test message\n" +
+                "--" + BOUNDARY + "--").getBytes());
+
+        DefaultFullHttpRequest req = new DefaultFullHttpRequest(HttpVersion.HTTP_1_0, HttpMethod.POST, "/up", byteBuf);
+        req.headers().add(HttpHeaderNames.CONTENT_TYPE, "multipart/form-data; boundary=" + BOUNDARY);
+
+        HttpPostRequestDecoder decoder =
+                new HttpPostRequestDecoder(new DefaultHttpDataFactory(DefaultHttpDataFactory.MINSIZE),
+                        req,
+                        java.nio.charset.Charset.forName("utf-8"));
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/HttpPostRequestDecoderTest.java]: use `CharsetUtil.UTF_8`

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/HttpPostRequestDecoderTest.java [hunk 3]
```diff
@@ -656,4 +656,36 @@ public void testDecodeMalformedEmptyContentTypeFieldParameters() throws Exceptio
...
+                "Content-Disposition: form-data; name=\"msg_id\"\n" +
+                "\n" +
+                "15200\n" +
+                "--" + BOUNDARY + "\n" +
+                "Content-Disposition: form-data; name=\"msg\"\n" +
+                "\n" +
+                "test message\n" +
+                "--" + BOUNDARY + "--").getBytes());
+
+        DefaultFullHttpRequest req = new DefaultFullHttpRequest(HttpVersion.HTTP_1_0, HttpMethod.POST, "/up", byteBuf);
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/HttpPostRequestDecoderTest.java]: nit: `FullHttpRequest req = ...`

**Comment by ssserj** [comment]
@normanmaurer ICLA was signed. Changes was made. But `req.refCnt()` is not 0. I will be research why?((

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/HttpPostRequestDecoderTest.java (hunk 1 above)
Comment (ssserj) [reply codec-http/src/test/java/io/netty/handler/codec/http/multipart/HttpPostRequestDecoderTest.java]: Most probably, should be assertEquals(1, req.refCnt()) ?

**Comment by normanmaurer** [comment]
@ssserj hmm... let me investigate a bit. I thought it should be 0 as we basically destroyed the decoder + everything that is still stored in there.

**Comment by normanmaurer** [comment]
@ssserj actually yes your are right 1 is correct as this class does not take over ownership as of today.

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/HttpPostRequestDecoderTest.java [hunk 4]
```diff
@@ -20,16 +20,7 @@
...
-import io.netty.handler.codec.http.DefaultHttpContent;
-import io.netty.handler.codec.http.DefaultHttpRequest;
-import io.netty.handler.codec.http.DefaultLastHttpContent;
-import io.netty.handler.codec.http.HttpContent;
-import io.netty.handler.codec.http.HttpHeaderNames;
-import io.netty.handler.codec.http.HttpHeaderValues;
-import io.netty.handler.codec.http.HttpMethod;
-import io.netty.handler.codec.http.HttpVersion;
-import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.handler.codec.http.*;
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/HttpPostRequestDecoderTest.java]: nit: please revert... we usually not do wildcard imports for this kind of things

**Comment by normanmaurer** [comment]
Squashed, cleaned up and cherry-picked as 9e2922b04d2901647b717232ed8fa1745bdbe0df into 4.1

**Pull request closed without merging**